core/src/forwarding_stage.rs

```diff
@@ -0,0 +1,618 @@
+    solana_cost_model::cost_model::CostModel,
+    solana_gossip::contact_info::Protocol,
+    solana_net_utils::bind_to_unspecified,
+    solana_perf::{data_budget::DataBudget, packet::Packet},
+    solana_poh::poh_recorder::PohRecorder,
+    solana_runtime::{bank::Bank, root_bank_cache::RootBankCache},
+    solana_runtime_transaction::{
+        runtime_transaction::RuntimeTransaction, transaction_meta::StaticMeta,
+    },
+    solana_sdk::{fee::FeeBudgetLimits, packet, transaction::MessageHash},
+    solana_streamer::sendmmsg::batch_send,
+    std::{
+        net::{SocketAddr, UdpSocket},
+        sync::{Arc, RwLock},
+        thread::{Builder, JoinHandle},
+        time::{Duration, Instant},
+    },
+};
+
+const FORWARD_BATCH_SIZE: usize = 128;
+
+#[derive(Clone)]
+pub struct ForwardingAddresses {
+    pub tpu: Option<SocketAddr>,
+    pub tpu_vote: Option<SocketAddr>,
+}
+
+pub trait ForwardAddressGetter: Send + Sync + 'static {
+    fn get_forwarding_addresses(&self, protocol: Protocol) -> ForwardingAddresses;
+}
+
+impl<T: LikeClusterInfo> ForwardAddressGetter for (T, Arc<RwLock<PohRecorder>>) {
+    fn get_forwarding_addresses(&self, protocol: Protocol) -> ForwardingAddresses {
+        ForwardingAddresses {
+            tpu: next_leader(&self.0, &self.1, |node| node.tpu_forwards(protocol)).map(|(_, s)| s),
+            tpu_vote: next_leader_tpu_vote(&self.0, &self.1).map(|(_, s)| s),
+        }
+    }
+}
+
+pub struct ForwardingStage<F: ForwardAddressGetter> {
+    receiver: Receiver<(BankingPacketBatch, bool)>,
+    packet_container: PacketContainer,
+
+    root_bank_cache: RootBankCache,
+    forward_address_getter: F,
+    connection_cache: Arc<ConnectionCache>,
+    data_budget: DataBudget,
+    udp_socket: UdpSocket,
+
+    metrics: ForwardingStageMetrics,
+}
+
+impl<F: ForwardAddressGetter> ForwardingStage<F> {
+    pub fn spawn(
+        receiver: Receiver<(BankingPacketBatch, bool)>,
+        connection_cache: Arc<ConnectionCache>,
+        root_bank_cache: RootBankCache,
+        forward_address_getter: F,
+    ) -> JoinHandle<()> {
+        let forwarding_stage = Self::new(
+            receiver,
+            connection_cache,
+            root_bank_cache,
+            forward_address_getter,
+        );
+        Builder::new()
+            .name("solFwdStage".to_string())
+            .spawn(move || forwarding_stage.run())
+            .unwrap()
+    }
+
+    fn new(
+        receiver: Receiver<(BankingPacketBatch, bool)>,
+        connection_cache: Arc<ConnectionCache>,
+        root_bank_cache: RootBankCache,
+        forward_address_getter: F,
+    ) -> Self {
+        Self {
+            receiver,
+            packet_container: PacketContainer::with_capacity(4 * 4096),
+            root_bank_cache,
+            forward_address_getter,
+            connection_cache,
+            data_budget: DataBudget::default(),
+            udp_socket: bind_to_unspecified().unwrap(),
```

Comment:
I think we should stop binding on all interfaces   @alexpyattaev

Reply:
what should I use instead? I based the actual forwarding code in here on how we previously did it

Reply:
Yeah, I know that  we do this way but we should stop doing this. Maybe Alex has in mind some opinion about how what function to use for this. I think that it would make sense to pass udp_socket address to the constructor, so that on the level of validator the bind address is selected, which by default should be `127.0.0.1`

Reply:
For the stuff in agave you should prefer functions bind_in_range or bind_in_range_with_config. This will allow your code to respect bind-address CLI argument. Not respecting it will break firewall policies for some users.

Reply:
Maybe I will abstract out all the network logic in this component, not sure, rebasing my old changes